jsk_perception/src/template_match_detector.cpp

```diff
@@ -0,0 +1,478 @@
+        img_pub_.publish(cv_bridge::CvImage(
+                                            image_msg->header,
+                                            sensor_msgs::image_encodings::BGR8,
+                                            result_img).toImageMsg());
+        return;
+      }
+    }
+
+    if(rects.size() > max_detect_num){
+      std::cout << "detected rectangles num is over maxdetected num. detected num : " << rects.size() << "  max num : "
+                << max_detect_num << std::endl;
+      if(specify_target_){
+        std::cout << "Increasing threshold before : " << matching_thre;
+        matching_thre = std::min(matching_thre + thre_update_step, 1.0);
+        std::cout << "  after : " << matching_thre << std::endl;
+        return;
+      } else {
+        // todo clear rects
+        return;
+      }
+    }
+
+    // integrate neighborhood rectangles
+    std::vector<cv::Rect> result_rects;
+    TemplateMatchDetector::rectangleIntegration(rects, score, result_rects);
+    if(screen_debug_)
+      std::cout << "integrated_rect num : " << result_rects.size() << std::endl;
+
+    //update matching threshold
+    if(specify_target_ &&
+       result_rects.size() != target_num){
+      if(result_rects.size() > target_num){
+        std::cout << "Too many rectangles detected. detected num : " << result_rects.size() << "  target num : "
+                  << target_num << std::endl << "Increasing threshold before : " << matching_thre;
+        matching_thre = std::min(matching_thre + thre_update_step, 1.0);
+        std::cout << "  after : " << matching_thre << std::endl;
+      }
+      else {
+        std::cout << "Too few rectangles detected. detected num : " << result_rects.size() << "  target num : "
+                  << target_num << std::endl << "Decreasing threshold before : " << matching_thre;
+        matching_thre = std::max(matching_thre - thre_update_step, 0.0);
+        std::cout << "  after : " << matching_thre << std::endl;
+      }
+      return;
+    }
+    if(screen_debug_) std::cout << std::endl;
+
+    //draw rectangles on image
+    for (i = 0;i < result_rects.size(); i++) {
+      cv::rectangle(result_img, result_rects.at(i).tl(), result_rects.at(i).br(), cv::Scalar(0, 0, 255), 2, 8, 0);
+    }
+
+    jsk_recognition_msgs::RectArray rects_msg;
+    rects_msg.rects.resize(result_rects.size());
+    for(i = 0; i < result_rects.size(); i++){
+      jsk_recognition_msgs::Rect rect;
+      rect.x = result_rects[i].x;
+      rect.y = result_rects[i].y;
+      rect.width = result_rects[i].width;
+      rect.height = result_rects[i].height;
+      rects_msg.rects[i] = rect;
+    }
+
+    rects_msg.header = image_msg->header;
+    rects_msg.rects.resize(result_rects.size());
+    rects_pub_.publish(rects_msg);
+    img_pub_.publish(cv_bridge::CvImage(
+                     image_msg->header,
+                     sensor_msgs::image_encodings::BGR8,
+                     result_img).toImageMsg());
```

Comment:
What about `img_pub_.getNumSubscribers()` to decide to make and publish `result_img`?

Reply:
And maybe visualizing rects is not so difficult.
You can use `draw_rects`: https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_perception/src/draw_rects.cpp#L63-L64,
so IMO visualizing in this node is not so good idea.

Reply:
> What about img_pub_.getNumSubscribers() to decide to make and publish result_img?

That's good idea. I added it.

> And maybe visualizing rects is not so difficult.
> You can use draw_rects: https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_perception/src/draw_rects.cpp#L63-L64,
> so IMO visualizing in this node is not so good idea.

Hmm, I think your idea is good. But light user could not catch up with a that's idea.